### ISSUE TITLE
Exceptions not propagated to `rejectionHandler`

### ISSUE THREAD

**Issue by stanch** [hseeberger/akka-http-json#46]
Hi,

I’m using the Play JSON module. If unmarshalling fails, it throws a `JsResultException`, which results in a `MalformedRequestContentRejection`. However the rejection only contains the stringified message of the exception, e.g. the object available to a rejection handler is something like

```scala
MalformedRequestContentRejection(
  "JsResultException(errors:List((/foo,List(ValidationError(List(error.path.missing),WrappedArray())))))",
  None
)
```

This prevents one from using the information contained in the exception in any meaningful way, e.g. to construct a pretty response for the client. The reason that the information is not available is [this code in Akka](https://github.com/akka/akka/blob/master/akka-http/src/main/scala/akka/http/scaladsl/server/directives/MarshallingDirectives.scala#L37-L38): 

```scala
case Failure(x: IllegalArgumentException) ⇒ reject(ValidationRejection(x.getMessage.nullAsEmpty, Some(x)))
case Failure(x) ⇒ reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, Option(x.getCause)))
```

So basically there are three ways to make the actual list of validation errors available in a rejection handler: either `JsResultException` should be wrapped in some other exception (so that it’s extracted by `x.getCause`), or a different exception subclassing `IllegalArgumentException` should be thrown, or both. Here’s an approach that worked for me:

In [`PlayJsonSupport`](https://github.com/hseeberger/akka-http-json/blob/master/akka-http-play-json/src/main/scala/de/heikoseeberger/akkahttpplayjson/PlayJsonSupport.scala#L38):

```scala
def read(json: JsValue) = reads.reads(json).recoverTotal { error =>
  throw new IllegalArgumentException { initCause(JsResultException(error.errors)) }
}
```

In a rejection handler:

```scala
implicit val rejectionHander = RejectionHandler.newBuilder()
  .handle { case ValidationRejection(_, Some(e)) ⇒  /* `e.getCause` is the `JsResultException` */ ??? }
  .result()
```

Do you think this change in `PlayJsonSupport` makes sense? If so, something similar needs to be done with `Json.parse(input)` to allow handling parsing errors. I haven’t used the other modules, but I imagine they can benefit from similar changes. If this makes sense (rather than, say, changing the behavior of unmarshalling directives themselves), I can submit a PR with these changes, at least for Play JSON.

**Comment by hseeberger** [comment]
@stanch, thanks for this very informative explanation. Rather than fixing all the modules I'd like to understand why the Akka HTTP code deals with other failures the way it does. Maybe Akka HTTP should create a `MalformedRequestContentRejection` using `x` instead of `x.getCause`. Could you please create a respective ticket against Akka, link this ticket and put me on cc?

**Comment by stanch** [comment]
I agree that it’s something that makes more sense in the long run. Will do that.

**Comment by stanch** [comment]
Fixed in Akka `2.4.5`.
